crates/pathfinder/src/rpc/v02/method/estimate_fee.rs

```diff
@@ -0,0 +1,418 @@
+use std::sync::Arc;
+
+use serde::Serialize;
+use serde_with::serde_as;
+
+use crate::{
+    cairo::ext_py::{BlockHashNumberOrLatest, GasPriceSource},
+    core::{BlockId, TransactionVersion},
+    rpc::v02::types::request::BroadcastedTransaction,
+    rpc::v02::{types::request::BroadcastedInvokeTransaction, RpcContext},
+    state::PendingData,
+};
+
+#[derive(serde::Deserialize, Debug, PartialEq, Eq)]
+pub struct EstimateFeeInput {
+    request: BroadcastedTransaction,
+    block_id: BlockId,
+}
+
+crate::rpc::error::generate_rpc_error_subset!(
+    EstimateFeeError: BlockNotFound,
+    ContractNotFound,
+    ContractError,
+    InvalidMessageSelector,
+    InvalidCallData
+);
+
+impl From<crate::cairo::ext_py::CallFailure> for EstimateFeeError {
+    fn from(c: crate::cairo::ext_py::CallFailure) -> Self {
+        use crate::cairo::ext_py::CallFailure::*;
+        match c {
+            NoSuchBlock => Self::BlockNotFound,
+            NoSuchContract => Self::ContractNotFound,
+            InvalidEntryPoint => Self::InvalidMessageSelector,
+            ExecutionFailed(e) => Self::Internal(anyhow::anyhow!("Internal error: {}", e)),
+            // Intentionally hide the message under Internal
+            Internal(_) | Shutdown => Self::Internal(anyhow::anyhow!("Internal error")),
+        }
+    }
+}
+
+impl TryInto<crate::rpc::v01::types::request::Call> for BroadcastedTransaction {
+    type Error = EstimateFeeError;
+
+    fn try_into(self) -> Result<crate::rpc::v01::types::request::Call, Self::Error> {
+        match self {
+            BroadcastedTransaction::Declare(_) | BroadcastedTransaction::Deploy(_) => {
+                Err(EstimateFeeError::Internal(anyhow::anyhow!(
+                    "Internal error: Only invoke version 0 transactions are supported."
```

Comment:
~~I'm a bit lost here - but one of the 2 cases that follow is v1 invoke txn.~~
EDIT: ok, the tests illustrate that clearly.

Reply:
No, you're right, the error message was incorrect.